File: sdk/go-guest/examples/http.go
package export_tests_http_api

import (
	wit "go.bytecodealliance.org/pkg/wit/types"
	http "wit_component/lsf/http"
)

func Run(which uint32, url string, _ uint64) uint64 {
	method := http.MethodPost
	if which == 0 {
		method = http.MethodGet
	}
	if which == 1 {
		method = http.MethodHead
	}
	r := http.Send(http.Request{Method: method, Url: url, Headers: []http.Header{},
		Body: wit.Some([]byte("payload")), BodyMediaType: wit.Some("text/plain"),
		IdempotencyKey: wit.None[string](), TimeoutMillis: wit.Some[uint64](1000)})
	if r.IsOk() {
		return uint64(r.Ok().Status) + 1000*uint64(len(r.Ok().Body))
	}
	switch r.Err().Tag() {
	case http.HttpErrorPermissionDenied:
		return 10
	case http.HttpErrorUncertain:
		return 11
	default:
		panic("unexpected HTTP outcome")
	}
}

File: sdk/typescript-guest/examples/http.ts
import type * as Contract from '../generated/interfaces/tests-http-api.js';
import * as http from '../vendor/lsf/sdk/typescript-guest/capabilities/http.js';
import { unwrap } from '../vendor/lsf/sdk/typescript-guest/capabilities/result.js';
export const api: typeof Contract = {
  run(which, url) {
    const result = http.send({ method: which === 0 ? 'get' : which === 1 ? 'head' : 'post', url,
      headers: [], body: new Uint8Array([112, 97, 121, 108, 111, 97, 100]),
      bodyMediaType: 'text/plain', timeoutMillis: 1000n });
    if (result.tag === 'err' && result.val.tag === 'permission-denied') return 10n;
    if (result.tag === 'err' && result.val.tag === 'uncertain') return 11n;
    const response = unwrap(result);
    return BigInt(response.status) + 1000n * BigInt(response.body.length);
  },
};
